### PR TITLE
Create a standalone logic to determine version status flags

### PR DESCRIPTION
**Pull request by suhaibmujahid**
This PR addresses the first three tasks in #2152.

## Checklist

- [x] Type annotations added to new functions
- [x] Docs added to functions touched in main classes
- [x] Dry-run produced the expected results
- [ ] The [`to-be-announced`](https://github.com/mozilla/bugbot/labels/to-be-announced) tag added if this is worth announcing

### CODE REVIEW

File: bugbot/rules/regression_set_status_flags.py [hunk 1]
```diff
@@ -204,6 +207,45 @@ def get_status_changes(self, bugs):
...
+            {
+                **bug,
+                "regressed_by": [bug["regressed_by"]],
+            }
+            for bug in bugs.values()
+        )
+        bugs_store = BugsStore(
+            chain(adjusted_bugs, data.values()), utils.get_checked_versions()
+        )
+        for bug_id, old_code_updates in self.status_changes.items():
```
Comment (marco-c) [inline bugbot/rules/regression_set_status_flags.py:225]: This is only checking that old updates are matched by new. We are not checking if there are new ones that were not caught before

File: bugbot/bug/analyzer.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        Args:
+            field: The field name.
+
+        Returns:
+            The field value. If the field is not found, `None` is returned.
+        """
+        return self._bug.get(field)
+
+    def detect_version_status_updates(self) -> list[VersionStatus] | None:
+        """Detect the status for the version flags that should to be updated.
```
Comment (marco-c) [inline bugbot/bug/analyzer.py:95]:
```suggestion
        """Detect the status for the version flags that should be updated.
```

File: bugbot/bug/analyzer.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+        The status of the version flags is determined by the status of the
+        regressor bug.
+
+        Returns:
+            A list of `VersionStatus` objects.
+        """
+        if len(self._bug["regressed_by"]) > 1:
+            # Currently only bugs with one regressor are supported
```
Comment (marco-c) [inline bugbot/bug/analyzer.py:104]: Could you file a follow-up to expand this support? When the regressors are all fixed in the same version, it's obvious. When they are fixed in different versions, we could set the latest as "affected" and the earliest as "?"

File: bugbot/bug/analyzer.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+            if fixed_version and fixed_version <= version:
+                # Bug was fixed in an earlier version, don't set the flag
+                continue
+            if (
+                version >= regressed_version
+                # ESR: If the regressor was uplifted, so the regression affects
+                # this version.
+                or regressor_bug.get_field(flag) in ("fixed", "verified")
+            ):
+                if is_latest_wontfix:
```
Comment (marco-c) [inline bugbot/bug/analyzer.py:141]: Why checking the latest version and not the `version` referring to the current flag being analyzed?

File: bugbot/bug/analyzer.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+                flag_updates.append(VersionStatus(channel, version, "unaffected"))
+
+        return flag_updates
+
+
+class BugNotInStoreError(LookupError):
+    """The bug was not found the bugs store."""
+
+
+class BugsStore:
```
Comment (marco-c) [inline bugbot/bug/analyzer.py:155]: What do you envision to be the long term usage for this class?

Comment (marco-c) [reply bugbot/bug/analyzer.py:155]: I'm wondering if it's worth having it or if we should just implement this stuff as part of the script itself

File: bugbot/rules/regression_set_status_flags.py (hunk 1 above)
Comment (suhaibmujahid) [reply bugbot/rules/regression_set_status_flags.py]: Nice catch! Yes, this is true. Fixed in 765e1fe65a7b5f778cc9dcf840c4018455e1b493.

File: bugbot/bug/analyzer.py (hunk 5 above)
Comment (marco-c) [reply bugbot/bug/analyzer.py:155]: All right, sounds good. I wonder if we can use similar logic to also merge actions from different rules.

**Review by marco-c — APPROVED** [review]

**Pull request merged**